content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,136 @@
+---
+title: "Good practices with NumPy random number generators"
+date: 2023-08-31T23:22:46+02:00
+draft: false
+description: "
+Good practices with NumPy random number generators
+"
+tags: ["tutorials", numpy, rng]
+displayInList: true
+author: ["Albert Thomas"]
+---
+
+Unless you are working on a problem where you can afford a true Random Number Generator (RNG), which is basically never for most of us, implementing something random means relying on a pseudo Random Number Generator. I want to share here what I have learned about good practices with pseudo RNGs and especially the ones available in [NumPy](https://numpy.org/). I assume a certain knowledge of NumPy and that NumPy 1.17 or greater is used. The reason for this is that great new features were introduced in the [random](https://numpy.org/doc/stable/reference/random/index.html) module of version 1.17. As `numpy` is usually imported as `np`, I will sometimes use `np` instead of `numpy`. Finally, as I will not talk about true RNGs, RNG will always mean pseudo RNG in the rest of this blog post.
+
+### The main messages
+1. Avoid using the global NumPy RNG. This means that you should avoid using [`np.random.seed`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html) and `np.random.*` functions, such as `np.random.random`, to generate random values.
+2. Create a new RNG and pass it around using the [`np.random.default_rng`](https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.default_rng) function.
+3. Be careful with parallel computations and rely on [NumPy strategies for reproducible parallel number generation](https://numpy.org/doc/stable/reference/random/parallel.html).
+
+Note that with NumPy <1.17 the way to create a new RNG is to use [`np.random.RandomState`](https://numpy.org/doc/stable/reference/random/legacy.html#numpy.random.RandomState) which is based on the popular Mersenne Twister 19937 algorithm. This is also how the global NumPy RNG is created. It is still possible to use this function in versions higher than 1.17 but it is now recommended to use `default_rng` which returns an instance of the statistically better [PCG64](https://www.pcg-random.org) RNG. You might see `np.random.RandomState` being used in tests as it has strong stability guarantees between different NumPy versions.
+
+## Random number generation with NumPy
+When you import `numpy` in your python script a RNG is created behind the scenes. This RNG is the one used when you generate a new random value using a function such as `np.random.random`. I will here refer to this RNG as the global NumPy RNG.
```

Comment:
```suggestion
When you import `numpy` in your python script, a RNG is created behind the scenes. This RNG is the one used when you generate a new random value using a function such as `np.random.random`. I will here refer to this RNG as the global NumPy RNG.
```

I would probably use "an RNG", but "a RNG" is OK if you imagine people reading "a random number generator".